Add shard parsing to  `nodes_stats_parser.py`

nervous-inhuman opened the pull request:
We want to be able to track which shards are on which nodes. 

This PR allows the exporter to parse the output about shards, 
when you've specified the 
`--nodes-stats-(fields|level shards|index-metrics)` parameters to output info about shards.

oneohtrix committed:
Add shard parsing to  `nodes_stats_parser.py`

oneohtrix committed:
Fix shard parsing

jbrunclik commented on prometheus_es_exporter/nodes_stats_parser.py:
```diff
@@ -37,7 +46,16 @@ def parse_block(block, metric=[], labels={}):
                     for n_key, n_value in value.items():
-                        result.extend(parse_block(n_value, metric=metric + [key], labels=merge_dicts(labels, {singular_key: [n_key]})))
+                        if isinstance(n_value, list):
+                            if key == "shards":
```
Please don't mix ```s and `"`s - see [here](https://github.com/google/styleguide/blob/gh-pages/pyguide.md#310-strings) for more details.

jbrunclik commented on prometheus_es_exporter/nodes_stats_parser.py:
```diff
@@ -37,7 +46,16 @@ def parse_block(block, metric=[], labels={}):
+                                for n_value in n_list:
+                                    result.extend(parse_block(n_value, metric=metric + [key], labels=merge_dicts(labels, {singular_key: [n_key]})))
+                        else:
+                            if key in excluded_metric_keys:
```
Why is the key being added to the result when it's supposed to be `excluded`? This seems semantically wrong.

nervous-inhuman replied:
The `excluded_metric_keys` specifies which keys shouldn't appear in the metric name, not that they shouldn't be added to the result. 

Since the metric name is built by iterating over the dict keys, I needed a way to exclude some specific keys (e.g. `shard_id`, in our case). That's why the parameter `metric` is set to itself, instead of `metric + [key]`.

If it seems confusing to you, would a comment that explains what it's used for be enough?

jbrunclik replied:
Yep, a comment would be enough for me. Thank you.

jbrunclik commented on prometheus_es_exporter/nodes_stats_parser.py:
```diff
@@ -4,16 +4,25 @@
     'pools': 'pool',
     'collectors': 'collector',
     'buffer_pools': 'buffer_pool',
+    'shards': 'shard'
```
It looks like the convention within this file is to have a trailing comma for the last element in the list. If we want to push this to upstream, then I would suggest to stick to the convention.

jbrunclik reviewed (changes requested):
Please provide a description for this PR explaining what is your intention with this PR.

oneohtrix committed:
Fixed stylistic issues; Changed shard label name

I think using `index` is more fitting, since we're displaying the indexes name. 

This is how the output for the shard metric looks like now:
`es_nodes_stats_indices_shards_routing_primary{node_id="0EWUhXeBQtaVGlexUeVwMg",node_name="node1_example_com",shard="logs-2018_10_07",shard_id="7"} 0.0`

jbrunclik commented on prometheus_es_exporter/nodes_stats_parser.py:
```diff
@@ -37,7 +46,16 @@ def parse_block(block, metric=[], labels={}):
-                        result.extend(parse_block(n_value, metric=metric + [key], labels=merge_dicts(labels, {singular_key: [n_key]})))
+                        if isinstance(n_value, list):
+                            if key == 'shards':
+                                n_list = [{"shard_id": d} for d in n_value]
```
Please replace `"` with `'` for consistency also here.

jbrunclik reviewed (commented):
Thanks for updating the description!

oneohtrix committed:
Stylistic changes; Comment for `excluded_metric_keys`
